### PR TITLE
sysadmin page locators actions and step definition addition

### PR DESCRIPTION
**Pull request by itsmekumari**
sysadmin page locators actions and step definition addition

### CODE REVIEW

File: src/main/java/stepsdesign/SysadminSteps.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  @Then("Enter the Namespace Description with value: {string}")
+  public void enterNamespaceDescriptionValue(String value) {
+    CdfSysAdminActions.enterNamespaceDescription(value);
+  }
+
+  @Then("Verify the Create a Profile page is loaded for selected provisioner")
+  public void verifyTheCreateAProfilePageIsLoaded() {
+    CdfSysAdminActions.verifyCreateProfilePageLoaded();
+  }
+
```
Comment (Vipinofficial11) [inline src/main/java/stepsdesign/SysadminSteps.java:151]: Remove the space

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java [hunk 2]
```diff
@@ -77,4 +84,237 @@ public static void verifySuccess() {
     }
     Assert.assertTrue(checkParam);
   }
+
+  /**
+   * Click on the System Admin Menu
+   */
+  public static void clickOnSystemAdminMenu(String menuName) {
+    ElementHelper.clickOnElement(CdfSysAdminLocators.locateMenu(menuName));
+  }
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java:95]: Few things here,

1. If we are simply using this method to select the System menu every time we don't have to pass the parameter.
2. We shouldn't actually make this method or Step name specfic to System Admin, lets make it re-usable.

Changes I would suggest.
1. Change the method name it shouldn't be page specific if method is common.
2.Update the step, it should accept one parameter just the definition will be a bit changed.

ex : ``` Open "System Admin" menu``` or ``` Open "Hub" page *change*```

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java [hunk 3]
```diff
@@ -77,4 +84,237 @@ public static void verifySuccess() {
...
+
+  /**
+   * Select the type of preferenceName to open in system admin
+   *
+   * @param preferenceName @data-cy attribute value of preference name. If preferenceName is present
+   *                       in {@link ConstantsUtil#DEFAULT_DATACY_ATTRIBUTES_FILE} then its data-cy
+   *                       is fetched from it else preferenceName is used as it is.
+   */
+  public static void clickOnHeading(String preferenceName) {
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java]: Can we change the method name to something more specific.

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java [hunk 4]
```diff
@@ -77,4 +84,237 @@ public static void verifySuccess() {
...
+   *
+   * @param preferenceProperty @data-cy attribute value of preference Property. If
+   *                           preferenceProperty is present in {@link ConstantsUtil#DEFAULT_DATACY_ATTRIBUTES_FILE}
+   *                           then its data-cy is fetched from it else preferenceProperty is used
+   *                           as it is.
+   * @param jsonKeyValuePairs  Actual json KeyValue Pairs string is fetched from {@link
+   *                           ConstantsUtil#DEFAULT_PLUGIN_PROPERTIES_FILE} with jsonKeyValuePairs
+   *                           as key
+   */
+  public static void enterKeyValuePreferences(String preferenceProperty, String jsonKeyValuePairs) {
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java]: ```jsonKeyValuePairs``` to ```keyValuePair```

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java [hunk 5]
```diff
@@ -77,4 +84,237 @@ public static void verifySuccess() {
...
+  }
+
+  /**
+   * Select on the type of Provisioner from list for Compute Profile in system admin
+   *
+   * @param provisionerName @data-cy attribute value of Provisioner. If Provisioner is present in
+   *                        {@link ConstantsUtil#DEFAULT_DATACY_ATTRIBUTES_FILE} then its data-cy is
+   *                        fetched from it else Provisioner is used as it is.
+   */
+  public static void selectProvisionerComputeProfile(String provisionerName) {
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java]: ```selectProvisionerComputeProfile``` to ```selectProvisioner```

File: src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java [hunk 6]
```diff
@@ -51,4 +53,92 @@ public class CdfSysAdminLocators {
...
+  public static WebElement createNewNamespace;
+  @FindBy(how = How.XPATH, using = "//button[@data-testid='edit-system-prefs-btn']")
+  public static WebElement editSystemPreference;
+  @FindBy(how = How.XPATH, using = "//a[@href='/cdap/ns/system/profiles/create']")
+  public static WebElement createComputeProfile;
+  @FindBy(how = How.XPATH, using = "//div//label[@for='import-profile']")
+  public static WebElement clickImport;
+  @FindBy(how = How.XPATH, using = "//button[@data-testid='save-prefs-btn']")
+  public static WebElement clickSaveClose;
+  @FindBy(how = How.XPATH, using = "//input[@placeholder='Namespace name' and @type='text']")
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java]: We have a data-test-id for this we can use it and create something like this :

```//div[@data-testid='create-namespace-name']//input```

File: src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java [hunk 7]
```diff
@@ -51,4 +53,92 @@ public class CdfSysAdminLocators {
...
+  public static WebElement clickDelete;
+  @FindBy(how = How.XPATH, using = "//button[@data-testid='Reload']")
+  public static WebElement clickReload;
+  @FindBy(how = How.XPATH, using = "//*[contains(text(),'Close')]")
+  public static WebElement closeButton;
+  @FindBy(how = How.XPATH, using = "//div[contains(@class, 'error-section text-danger')]")
+  public static WebElement errorMessageOnFooter;
+  @FindBy(how = How.XPATH, using = "//div[contains(@class, 'card-action-feedback DANGER')]")
+  public static WebElement failMessage;
+
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java:94]: A common comment for above locators, if we can find an Id for these locators we should use them. Avoid using such locator as much as possible. 

```@type='submit' and @class='btn remove-row-btn btn-link']```

Note - Class names are mostly defined by developers to apply styling/css on web pages and they can change them quite frequently while an UI update.

File: src/main/java/stepsdesign/SysadminSteps.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    CdfSysAdminActions.clickOnHeading(pageName);
+  }
+
+  @Then("Click on edit system preferences to set system preferences")
+  public void clickOnEditSystemPreferences() {
+    CdfSysAdminActions.clickOnEditPreference();
+  }
+
+  @Then("Set system preferences with key: {string} and value: {string}")
+  public void setSystemPreferencesWithKeyAndValue(String pluginProperty, String jsonKeyValuePairs) {
```
Comment (Vipinofficial11) [inline src/main/java/stepsdesign/SysadminSteps.java]: jsonKeyValuePairs -> KeyValuePairs

File: src/main/java/stepsdesign/SysadminSteps.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * CDF sysadmin related step design.
+ */
+public class SysadminSteps implements CdfHelper {
+
+  private static final Logger logger = LoggerFactory.getLogger(SysadminSteps.class);
+
+  @Then("Click on the {string} link from menu")
```
Comment (Vipinofficial11) [inline src/main/java/stepsdesign/SysadminSteps.java]: Overall the changes looks good for this page we just may have to revisit the step definition naming approach.

For example I see definition namings like
- Select the Namespace from the System admin page (Can be updated to ```Select the Namespace```)
- Click on edit system preferences to set system preferences (Can be updated to ``` Click on edit system preferences```)
- Click on Delete button to delete the added preference and Click on Reset button to reset the added preferences
``` Delete the preferences and Reset the preferences```

These are just for idea, Step definition should clearly state what the step is doing. We may not have to provide additional information.

Discuss this internally once with the QA team, based on the discussion you can make the changes if required.

**Comment by Vipinofficial11** [comment]
A common comment for the code indentation, use all 120 characters in a line and then use the new line or if want to separate the logic then use the new line.

ex : 

```
String xpath = "//*[@data-cy='" + pluginProperty + "" + row + "']//span
        //button[@type='submit' and @class='btn remove-row-btn btn-link']";
```

File: src/main/java/stepsdesign/SysadminSteps.java (hunk 1 above)
Comment (itsmekumari) [reply src/main/java/stepsdesign/SysadminSteps.java:151]: Removed space.

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java (hunk 2 above)
Comment (itsmekumari) [reply src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java:95]: Changed the step definition, parameter was already added.

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java (hunk 3 above)
Comment (itsmekumari) [reply src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java]: Change done.

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java (hunk 4 above)
Comment (itsmekumari) [reply src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java]: Change done.

File: src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java (hunk 5 above)
Comment (itsmekumari) [reply src/main/java/io/cdap/e2e/pages/actions/CdfSysAdminActions.java]: Change done

File: src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java (hunk 6 above)
Comment (itsmekumari) [reply src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java]: Change done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java (hunk 7 above)
Comment (itsmekumari) [reply src/main/java/io/cdap/e2e/pages/locators/CdfSysAdminLocators.java:94]: Changed for the locators which has Id, but for few they are not available.

File: src/main/java/stepsdesign/SysadminSteps.java (hunk 8 above)
Comment (itsmekumari) [reply src/main/java/stepsdesign/SysadminSteps.java]: Changed.

File: src/main/java/stepsdesign/SysadminSteps.java (hunk 9 above)
Comment (itsmekumari) [reply src/main/java/stepsdesign/SysadminSteps.java]: The Navigations are different from sysadmin and hence used system admin for first step, for others changed as suggested.

**Comment by itsmekumari** [comment]
> A common comment for the code indentation, use all 120 characters in a line and then use the new line or if want to separate the logic then use the new line.
> 
> ex :
> 
> ```
> String xpath = "//*[@data-cy='" + pluginProperty + "" + row + "']//span
>         //button[@type='submit' and @class='btn remove-row-btn btn-link']";
> ```


Used the code formatter intellij-java-google-style.xml which changed the format, changed as suggested.

**Review by bharatgulati — APPROVED** [review]

**Review by Vipinofficial11 — APPROVED** [review]

**Pull request merged**